### data/README.txt
### funvar-tracerx
**Nov '23**

Data folder for submission with datasets referenced by scripts in scripts folder.

**Note: AAindex amino acid substitution matrices provided externally by:**
`ftp://ftp.genome.jp/pub/db/community/aaindex/`
See `data/aaindex/README.txt`

---
`data/_yj13_tcga_only_count_fies_by_score_and_cgc.tsv`
	Counts FIEs by FunVar score and CGC genes for distribution plot

---
```
data/diversity_yj07f_full_table_gene.csv
data/diversity_yj07f_full_table_funfam.csv
```
FIE and mutations tables used to calculate Hill-Shannon diversity.

Generated on 20231101 by: `script/diversity_analysis/neofun_paper_counts_yj_diversity.sql	Â§A` 
Refer to: 
`script/diversity_analysis/neofun_paper_diversity_simplified.R`
`data/vw_diversity.tsv` An export of the db view vw_diversity for reference

---

`data/neofun_v3.2_st_yj04d_0526.tsv`
FIE/gene/tumour overall counts by cancer type

**Note**: 	
`SCORE_THRESHOLD=="MC3"` means FunVar scores >=3 as used in manuscript
`TXP == "N"`	refers to TRACERx-lung
`TXP == "Y"`	refers to TCGA-lung (post-processed via Tx pipeline)

--- 

`data/nfe_score_database_py_funvar_archive_nfe_pfh_mutfam_090_exported_no_mut_id.tsv`
Table of mutations, clusters, func sites etc used for calculating FIE scores generated by: 
`script/nfe_score_database_py_export.sql`

Refer to: 
`script/fie_scoring/nfe_main.py` and classes defined in `packages/mutation.py` and `neofunevent.py`

**Note**: neofun event (NFE) is old terminology for FIE. FIEs are a pool of functional impact events which *may* contain neofunctional events; evidence is presented in manuscript for possible cases.

